### PR TITLE
Refactor Intermediate format graph for supporting multiple backend

### PR DESCRIPTION
**Pull request by dboyliao**
IR based code generator

### CODE REVIEW

File: utensor_cgen/code_generator.py [hunk 1]
```diff
@@ -20,24 +19,35 @@
...
-               clear_tmp_file=True):
-    self.pb_file = pb_file
+               **trans_kwargs):
+    self.model_file = model_file
     if not os.path.exists(idx_dir):
       os.makedirs(idx_dir)
     self.idx_dir = idx_dir
     self.embed_data_dir = embed_data_dir.rstrip("/")
-    self.method = method
+    self.methods = methods
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/code_generator.py]: Um... the naming is not good honestly.
`methods` means the transformation methods you want to apply to your graph.
This will be used to setup a transformation pipeline later.

File: utensor_cgen/code_generator.py [hunk 2]
```diff
@@ -48,38 +58,40 @@ def generate(self, src_fname):
...
-    construct_order = RefCntOptimizer.optimize(ops_info, topo_orders, self.output_nodes, self.method)
-
-    # TODO better snippet construction abstraction
-    for op_id, (op_name, op_info, ref_counts, to_eval) in enumerate(construct_order, 1):
+    graph_def = tf.GraphDef()
+    with tf.gfile.FastGFile(self.model_file, 'rb') as fid:
+      graph_def.ParseFromString(fid.read())
+    ugraph = uTensorGraph(graph_def, self.output_nodes)
+    print("Transforming graph: {}".format(self.model_file))
+    quant_ugraph = self._transform_graph(ugraph,
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/code_generator.py]: I suppose there should be different quantization process for other graphs.
However, with `uTensorGraph`, I'm able to construct a `GraphDef` out of it.
So maybe we can quantize other graphs with Tensorflow quantization? Not sure.

File: utensor_cgen/ir/base.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
+# -*- coding: utf8 -*-
+import six
+from copy import deepcopy
+import re
+
+import attr
+from attr import validators
+import numpy as np
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Ok, a `try catch` here?

File: utensor_cgen/ir/base.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+  dtype = attr.ib(validator=validators.instance_of(np.dtype))
+  shape = attr.ib(validator=validators.instance_of((list, type(None))))
+  @shape.validator
+  def check(self, attrib, values):
+    if values is not None:
+      for v in values:
+        assert isinstance(v, (int, type(None))), \
+          "shape should be a list of integers"
+
+  def __deepcopy__(self, memo):
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Well, it's a python convention for `deepcopy`.
https://stackoverflow.com/questions/1500718/what-is-the-right-way-to-override-the-copy-deepcopy-operations-on-an-object-in-p

[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: `memo` is like a cache which you can use to get data you need to reconstruct a object.
I don't think cpp convention applies here.

File: utensor_cgen/ir/base.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+  """
+  name : str
+  dtype : numpy.dtype
+  shape : list
+  """
+  name = attr.ib(validator=validators.instance_of(six.text_type))
+  dtype = attr.ib(validator=validators.instance_of(np.dtype))
+  shape = attr.ib(validator=validators.instance_of((list, type(None))))
+  @shape.validator
+  def check(self, attrib, values):
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Naming `shape` --> `shape_values`?
But it can be `None` actually (go ask `tf` team why)

File: utensor_cgen/transformer/ns_transformer.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+# -*- coding:utf8 -*-
+r"""Namescope Transformer
+
+Transformers that get rid of namescope/nodes which are not needed
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/transformer/ns_transformer.py]: Well, I think my comment here is not accurate in deed.
An `NamescopeTransformer` is just transformer which knowns how to deal with `tf.namescope` things.
It can remove it, replace it with other operations,...etc.
They are still transformers and in my definition for transformer, it is an object which has a method `transform` which will take a `uTensorGraph` and return a new `uTensorGraph`.

I think what you propose is more appropriate for an optimizer, which is also an graph transformer.

File: utensor_cgen/code_generator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/code_generator.py]: Let's change it to `transform_methods` then for now

File: utensor_cgen/code_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/code_generator.py]: i think we should add a check for current quantization. 
```
if already_quantized(graph):
  print warning
  skip_quantization()
```

File: utensor_cgen/ir/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: not here, probably at the cli level

File: utensor_cgen/ir/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: `def check(self, attrib, shape_values):`

File: utensor_cgen/ir/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: kk works for me

File: utensor_cgen/operators.py [hunk 7]
```diff
@@ -13,161 +14,218 @@ def snippet(self):
...
+  def __init__(self, op_info):
     _Operator.__init__(self)
-    inputs = [tname for tname, _, _ in op_info.input_tensor]
-    outputs = [tname for tname, _, _ in op_info.output_tensor]
-    _, in_dtype, _ = op_info.input_tensor[0]
-    _, filter_dtype, _ = op_info.input_tensor[1]
-    out_dtypes = [out_dtype for _, out_dtype, _ in op_info.output_tensor]
-    strides = op_info.op_attr["strides"].list.i
-    padding = op_info.op_attr["padding"].s.decode("utf8")
+    inputs = [tname for tname, _, _ in op_info.input_tensors]
```
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/operators.py]: kk

File: utensor_cgen/snippets/_types.py [hunk 8]
```diff
@@ -2,14 +2,22 @@
...
 _TYPE_MAP_VALUE = namedtuple("_TYPE_MAP_VALUE", ["importer_type_str", "tensor_type_str"])
 
-TF_TYPES_MAP = {
-  tf.float32: _TYPE_MAP_VALUE(importer_type_str="float", tensor_type_str="float"),
-  tf.qint8: _TYPE_MAP_VALUE(importer_type_str="byte", tensor_type_str="uint8_t"),
-  tf.int32: _TYPE_MAP_VALUE(importer_type_str="int", tensor_type_str="int"),
-  tf.int64: _TYPE_MAP_VALUE(importer_type_str="int", tensor_type_str="int"),
-  tf.quint8: _TYPE_MAP_VALUE(importer_type_str="ubyte", tensor_type_str="uint8_t"),
-  tf.qint32: _TYPE_MAP_VALUE(importer_type_str="int", tensor_type_str="int")
+NP_TYPES_MAP = {
```
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/snippets/_types.py]: works for me. Might want to add docs somehwere saying how to add new values, etc

File: utensor_cgen/transformer/ns_transformer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/transformer/ns_transformer.py]: cool beans

File: utensor_cgen/code_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/code_generator.py]: Um.... Checking whether if there is node like `Quantize`/`Dequantize`?

File: utensor_cgen/ir/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Ok, can do that.

File: utensor_cgen/ir/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Ok, got it.

File: utensor_cgen/ir/base.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+
+@attr.s
+class OperationInfo(IRBase, _NoShallowCopyMixin):
+  """
+  name : str
+  input_tensors : List[TensorInfo]
+  output_tensors : List[TensorInfo]
+  op_type : str
+  backend : str
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: kk.

File: utensor_cgen/ir/base.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+  def check(self, attribute, value):
+    if not all([isinstance(v, TensorInfo) for v in value]):
+      raise ValueError('Expecting a list of TensorInfo for output_tensor')
+
+  op_type = attr.ib(validator=validators.instance_of(six.text_type))
+
+  backend = attr.ib(validator=validators.instance_of(six.text_type))
+  @backend.validator
+  def check(self, attribute, value):
+    if value not in ['tensorflow']:
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Good idea.
I'll add a property like `all_supported_backends` later

File: utensor_cgen/ir/base.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+      raise ValueError('Unsupported backend: {}'.format(value))
+
+  op_attr = attr.ib(factory=dict)
+
+  def __attrs_post_init__(self):
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Oh, it's `attrs` things.
Maybe you can google `attrs __attrs_post_init__` for what it does.
But I'll comment it later.

File: utensor_cgen/ir/base.py [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+class uTensorGraph(IRBase, _NoShallowCopyMixin):
+  """
+  Attributes
+  ==========
+  ops_info : dict
+  topo_order : list
+  output_nodes : list
+  backend : str
+  """
+  KWPARSER_PATTERN = re.compile(r'^([^\d\W][\w\d_]*)__([^\d\W][\w\d_]*)')
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: I forget what is `+?` means honestly.
What's the difference?
I'm terrible at `re` XD

File: utensor_cgen/ir/base.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+  
+  @property
+  def backend(self):
+    return self._backend
+
+  @property
+  def graph_def(self):
+    assert self._backend == 'tensorflow', \
+      'Convert a uTensorGraph to tf.GraphDef from a non-tf backend'
+    graph_def = tf.GraphDef()
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Well, this is for quantization.
For now, we have to removing dropout layer before quantization.
Removing dropout layer after quantization is really weird and hard to work with.
So I have to removing dropout layer from a `uTensorGraph`, make it a `GraphDef`, quantize it, then make it back as a `uTensorGraph`.
Tell me if you have a better solution....

File: utensor_cgen/ir/converter.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+  @_check_tf_type
+  def get_generic_value(cls, value):
+    dtype = _DType(value)
+    np_dtype = dtype.as_numpy_dtype()
+    return self._handle_qtype(np_dtype)
+  
+  def _handle_qtype(self, dtype):
+    """Mapping qantized dtype to np.uint8
+
+    FIXME: I'm not sure if it's a good idea
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/converter.py]: Nice

File: utensor_cgen/ir/converter.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+    """
+    if dtype.fields is None:
+      return dtype
+    if dtype[0] in [np.uint8]:
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/converter.py]: You can extend it later if needed. :)

File: utensor_cgen/operators.py [hunk 16]
```diff
@@ -13,161 +14,218 @@ def snippet(self):
...
+  def __init__(self, op_info):
     _Operator.__init__(self)
-    inputs = [tensor_info.name for tensor_info in op_info.input_tensor]
-    output = op_info.output_tensor[0].name
-    tf_dtype = op_info.input_tensor[0].dtype
-    ref_count = ref_counts[0]
+    inputs = [tensor_info.name for tensor_info in op_info.input_tensors]
+    output = op_info.output_tensors[0].name
+    tf_dtype = op_info.input_tensors[0].dtype
+    parser = NamescopedKWArgsParser(RefCntOptimizer.KWARGS_NAMESCOPE, op_info.op_attr)
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/operators.py]: Yes, I agree.
But I don't have good idea for how to better refactor cpp code generation process here.

File: utensor_cgen/ir/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Well, on a second though on it, so basically you are saying tf/pytorch..etc is not required for installing `utensor_cgen` then check later when the program run?
And when install `utensor_cgen`, the user use sth like `pip install utensor_cgen[tensorflow]` for tensorflow support and maybe `pip install utensor_cgen[pytorch]` for pytorch support?

File: utensor_cgen/ir/base.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+                         output_tensors=deepcopy(self.output_tensors, memo),
+                         op_type=self.op_type,
+                         backend=self.backend,
+                         op_attr=deepcopy(self.op_attr, memo))
+
+
+class uTensorGraph(IRBase, _NoShallowCopyMixin):
+  """
+  Attributes
+  ==========
```
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: I found this definition is not correct for a freezed graph.
Since all variables are also transformed into const tensors which are noes with no inputs, either.
This definition will mix everything up.

Make it required argument?

File: utensor_cgen/ir/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: I add a property called `all_supported_backends` in the `IRBase` class

File: utensor_cgen/code_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/code_generator.py]: yes, or just making sure all tensors are uint8 or uint16

File: utensor_cgen/ir/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: "Well, on a second though on it, so basically you are saying tf/pytorch..etc is not required for installing utensor_cgen then check later when the program run?" Yes

"the user use sth like pip install utensor_cgen[tensorflow] for tensorflow support and maybe pip install utensor_cgen[pytorch] for pytorch support?" No

We should be able to handle this with the structure of our codebase.

```python
if input_file is protobuf:
  try:
    import tensorflow as tf
  except ModuleNotFoundExeption
    log.error("Failed to load tensorflow, is it installed? You may need to run pip install tensorflow")
    sys.exit(-1)
```

File: utensor_cgen/ir/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: swag

File: utensor_cgen/ir/base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: cool thanks

File: utensor_cgen/ir/base.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: hmm i think we should leave it for now then

File: utensor_cgen/ir/base.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: lol eww

File: utensor_cgen/ir/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: i need to think on it a little more. Not completely convinced about adding another feature just yet

[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: OK, I propose we do this in the future PR but not this one.

[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Maybe open an issue?

File: utensor_cgen/code_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/code_generator.py]: Ok, I'll do it this night after work.

[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/code_generator.py]: Instead of skip quantization, I show up a warning message to the user.
Since the transformation is also configurable by the user, there is no trivial way to determine what a user want based on all possible transformation and given graph type.

It's the user's responsibility to do it right.
At least they can fix the issue by the warning message if any.

File: utensor_cgen/ir/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/ir/base.py]: Yeah, lets open an issue and have a group discussion there.

File: utensor_cgen/code_generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbartling) [reply utensor_cgen/code_generator.py]: ok, we might want to give them the option to skip quantization entirely as a flag.

File: utensor_cgen/ir/base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (dboyliao) [reply utensor_cgen/ir/base.py]: Ok, after a few googling, I think we should go for `+?`.
I will fix it later.
